crates/ruff_linter/src/rules/refurb/rules/hardcoded_string_charset.rs

```diff
@@ -0,0 +1,168 @@
+            bytes,
+            bitset,
+        }
+    }
+}
+
+const KNOWN_CHARSETS: [Charset; 9] = [
+    Charset::new(
+        "ascii_letters",
+        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ",
+    ),
+    Charset::new("ascii_lowercase", "abcdefghijklmnopqrstuvwxyz"),
+    Charset::new("ascii_uppercase", "ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
+    Charset::new("digits", "0123456789"),
+    Charset::new("hexdigits", "0123456789abcdefABCDEF"),
+    Charset::new("octdigits", "01234567"),
+    Charset::new("punctuation", "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"),
+    Charset::new(
+        "printable",
+        "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"\
+        #$%&'()*+,-./:;<=>?@[\\]^_`{|}~ \t\n\r\x0b\x0c",
+    ),
+    Charset::new("whitespace", " \t\n\r\x0b\x0c"),
+];
+
+fn check_charset_as_set(s: &str) -> Option<&Charset> {
+    if !s.is_ascii() {
+        return None;
+    }
+    let bitset = s
+        .as_bytes()
+        .iter()
+        .try_fold(0, |acc, &byte| (byte < 128).then(|| acc | (1_u128 << byte)))?;
+
+    KNOWN_CHARSETS
+        .iter()
+        .find(|&charset| charset.bitset == bitset)
+}
+
+fn check_charset_exact(bytes: &[u8]) -> Option<&Charset> {
+    KNOWN_CHARSETS
+        .iter()
+        .find(|&charset| charset.bytes == bytes)
+}
+
+fn push_diagnostic(checker: &mut Checker, range: TextRange, charset: &Charset) {
+    let name = charset.name;
+    let mut diagnostic = Diagnostic::new(HardcodedStringCharset { name }, range);
+    diagnostic.try_set_fix(|| {
+        let (edit, binding) = checker.importer().get_or_import_symbol(
+            &ImportRequest::import("string", name),
+            range.start(),
+            checker.semantic(),
+        )?;
+        Ok(Fix::safe_edits(
+            Edit::range_replacement(binding, range),
+            [edit],
+        ))
+    });
+    checker.diagnostics.push(diagnostic);
+}
+
+/// FURB156
+pub(crate) fn hardcoded_string_charset_comparison(
+    checker: &mut Checker,
+    ops: &[CmpOp],
+    comparators: &[Expr],
+) {
+    let ([op], [comp]) = (ops, comparators) else {
+        return;
+    };
+    if !matches!(op, CmpOp::In | CmpOp::NotIn) {
+        return;
+    }
+    let Expr::StringLiteral(ExprStringLiteral { value, .. }) = comp else {
+        return;
+    };
+    let Some(charset) = check_charset_as_set(value.to_str()) else {
+        return;
+    };
+
+    // In this case the diagnostic will be emitted via string_like check.
+    if check_charset_exact(value.to_str().as_bytes()).is_some() {
+        return;
+    }
+
+    push_diagnostic(checker, comp.range(), charset);
+}
+
+/// FURB156
+pub(crate) fn hardcoded_string_charset_literal(checker: &mut Checker, string_like: StringLike) {
+    let StringLike::String(expr) = string_like else {
+        return;
+    };
+    let Some(charset) = check_charset_exact(expr.value.to_str().as_bytes()) else {
+        return;
+    };
+    push_diagnostic(checker, string_like.range(), charset);
+}
```

Comment:
You can move the invocation of this rule to the string-literal phase if it only needs to run on string literals.
```suggestion
pub(crate) fn hardcoded_string_charset_literal(checker: &mut Checker, string_literal: &ExprStringLiteral) {
    let StringLike::String(expr) = string_like else {
        return;
    };
    let Some(charset) = check_charset_exact(expr.value.to_str().as_bytes()) else {
        return;
    };
    push_diagnostic(checker, string_like.range(), charset);
}
```

 to https://github.com/astral-sh/ruff/blob/40149a17f5cf061664b849f7d6c0d03e42616038/crates/ruff_linter/src/checkers/ast/analyze/expression.rs#L1100-L1101

Reply:
IIUC, the right place was https://github.com/astral-sh/ruff/blob/a4a585da2a98f94407e602fc2328d5e4207185f3/crates/ruff_linter/src/checkers/ast/analyze/expression.rs#L1370
Moved.